include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,462 @@
+
+namespace boost { namespace geometry
+{
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail { namespace is_valid
+{
+
+template <typename PolyhedralSurface, bool AllowInconsistentOrientation = false>
+struct valid_intersection
+{
+    valid_intersection() = default;
+
+    using CT = typename coordinate_type<PolyhedralSurface>::type;
+    using vec3 = boost::qvm::vec<CT, 3>;
+    using polygon_3d = typename boost::range_value<PolyhedralSurface>::type;
+    using ring_3d = typename boost::geometry::ring_type<polygon_3d>::type;
+    using point_3d = typename boost::range_value<ring_3d>::type;
+    using segment_3d = typename boost::geometry::model::segment<point_3d>;
+    using set_vec_from_point = typename strategy::transform::detail::matrix_transformer::
+        set_vec_from_point<point_3d, 0, 3>;
+
+    using point_2d = boost::geometry::model::point
+        <CT, 2, boost::geometry::cs::cartesian>;
+    using linestring_2d = typename boost::geometry::model::linestring<point_2d>;
+    using polygon_2d = typename std::conditional
+        <
+            boost::geometry::point_order<polygon_3d>::value == boost::geometry::counterclockwise,
+            typename boost::geometry::model::polygon<point_2d, true>,
+            typename boost::geometry::model::polygon<point_2d, false>
+        >::type;
+    using ring_2d = typename boost::geometry::ring_type<polygon_2d>::type;
+    using segment_2d = typename boost::geometry::model::segment<point_2d>;
+
+private:
+
+    // Tolerance for points considered colinear, coplanar, etc.
+    constexpr static CT tolerance = std::numeric_limits<CT>::epsilon();
+
+    enum class intersection_type
+    {
+        none,
+        valid,
+        invalid
+    };
+
+    struct polygon_projection
+    {
+        polygon_2d projected_polygon;
+        vec3 base;
+        vec3 v1;
+        vec3 v2;
+        vec3 normal;
+    };
+
+    // Check if two vectors are linearly dependent (collinear)
+    static bool are_linearly_dependent(vec3 const& a, vec3 const& b)
+    {
+        return qvm::mag_sqr(qvm::cross(a, b)) < tolerance * tolerance;
+    }
+
+    // Try to find 3 non-collinear points
+    static bool find_non_collinear_points(polygon_3d const& polygon, vec3& base, vec3& v1, vec3& v2)
+    {
+        for (auto pit1 = points_begin(polygon); pit1 != points_end(polygon); ++pit1)
+        {
+            for (auto pit2 = pit1; pit2 != points_end(polygon); ++pit2)
+            {
+                for (auto pit3 = pit2; pit3 != points_end(polygon); ++pit3)
+                {
+                    set_vec_from_point::apply(*pit1, base);
+                    set_vec_from_point::apply(*pit2, v1);
+                    set_vec_from_point::apply(*pit3, v2);
+                    if (!are_linearly_dependent(v1 - base, v2 - base))
+                    {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false; // All points are collinear
+    }
```

Comment:
I think by the library design philosophy and also practically, if one wants to be able to plug in a robust computation later, this is a cartesian strategy.

Reply:
Indeed, a lot of computations here are coordinate system spesific and should be moved in strategies. On one hand, I think that polyhedral surface can only be cartesian. But on the other hand, predicates should be moved to stategies anyways for robustness reasons (as you said).

Reply:
That reduces the computation to orientation, great. I was thinking using the `project_point_to_plane_2d` to only do one orientation test but I think projecting to coordinate planes as you suggested is better since it does not involve any computations (substitutions, dot products etc). 
I change this part to use the side strategy provided by the umbrella strategy used in is_valid. This uses the side_by_triangle strategy as the default but the user can change it (as in many other cases in the library). At this point I think this is better then using a creating a new umbrella strategy that has side_rounded_input as a side strategy. We can think of substituting the side_by_triangle by side_rounded_input in general or in as many algorithms as possible.